turnstile-example/turnstile/examples/cs478/current/stlc.rkt

```diff
@@ -442,3 +446,23 @@
   #;  [⊢ (if- (inl-? e-)
           (let- ([x- (inl--val e-)]) e1-)
           (let- ([y- (inr--val e-)]) e2-)) ⇒ τ])
+
+(provide typed-match)
+(define-typerule typed-match
+  [(_ (x:id e) b) ≫
+   [⊢ e ≫ e- ⇒ τ]
+   [[x ≫ x- : τ] ⊢ b ≫ b- ⇒ τ-out]
+   ---------------------------------
+   [⊢ (let- ([x- e-]) b-) ⇒ τ-out]]
+  [(_ ([x:id ...] e) b) ≫
+   [⊢ e ≫ e- ⇒ (~× τ-tup ...)]
+   #:fail-unless (= (stx-length #'(x ...)) (stx-length #'(τ-tup ...)))
+   (format "~a pattern variables does not match ~a values in tuple" (stx-length #'(x ...)) (stx-length #'(τ-tup ...)))
```

Comment:
I'm still struggling with getting the error message right. Right now, a failed tuple pattern match falls through to the `rec` pattern match case (see the commented-out test case at the bottom of `stlc-prog.rkt`). How can I prevent that?